core/message.go

```diff
@@ -93,19 +92,37 @@ func (msg *Message) GetPrevStreamID() MessageStreamID {
 
 // GetRouter returns the stream object behind the current StreamID.
 func (msg *Message) GetRouter() Router {
-	return StreamRegistry.GetRouterOrFallback(msg.GetStreamID())
+	return StreamRegistry.GetRouterOrFallback(msg.streamID)
 }
 
 // GetPrevRouter returns the stream object behind the previous StreamID.
 func (msg *Message) GetPrevRouter() Router {
 	return StreamRegistry.GetRouterOrFallback(msg.prevStreamID)
 }
 
+// GetOrigRouter returns the stream object behind the original StreamID.
+func (msg *Message) GetOrigRouter() Router {
+	return StreamRegistry.GetRouterOrFallback(msg.origStreamID)
+}
+
 // SetStreamID sets a new stream and stores the current one in the previous
-// stream field.
+// stream field. If the previous stream is invalid and the new stream is not,
+// the new stream will become the original stream.
 func (msg *Message) SetStreamID(streamID MessageStreamID) {
-	msg.prevStreamID = msg.GetStreamID()
-	msg.data.streamID = streamID
+	if streamID != InvalidStreamID && msg.streamID == InvalidStreamID {
```

Comment:
I'm not sure if the following case is not possible: 

filter or formatter set stream id to `InvalidStreamID; router set new stream id to a fallback stream id => in this case original stream id will changed also to the fallback stream id.